### PR TITLE
Add NOTEARS algorithm to causal_discovery module

### LINKED ISSUE #1342: [ENH] NOTEARS Algorithm
### Subject of the issue
Requesting for adding a new structural learning algorithm (NOTEARS) for learning large networks.
This has been added to causalnex package.

Historically, structure learning has been a very hard problem. We are interested in looking for the optimal directed acyclic graph (DAGs) that describes the conditional dependencies between variables. However, the search space for this is combinatorial and scales super-exponentially with the number of nodes. NOTEARS algorithm cleverly introduces a new optimisation heuristic and approach to solving this problem, where the runtime for this is no longer exponential but scales cubically with the number of nodes.

References:
https://github.com/xunzheng/notears
https://causalnex.readthedocs.io/en/latest/05_resources/05_faq.html#why-notears-algorithm-over-other-structure-learning-methods

### Your environment
Enhancement request.

### Steps to reproduce
Enhancement Request.

### Expected behaviour
Enhancement request.

### Actual behaviour
Enhancement request.

### PR DESCRIPTION
**Pull request by GunaPalanivel**
## What

Adds NOTEARS as a structure learning algorithm under `pgmpy.causal_discovery`.

## Why

- Fixes #1342

## How

- Implemented NOTEARS using the `_BaseCausalDiscovery` API with sklearn style fit interface
- Formulated DAG learning as a continuous optimization problem with the matrix exponential acyclicity constraint
- Uses scipy L-BFGS-B for numpy backend
- Uses custom gradient descent for torch backend
- Supports l2 for continuous data, logistic for binary data, and poisson for count data
- ExpertKnowledge handling:

  - forbidden edges and temporal ordering handled as hard constraints using bounds or masking
  - required edges handled as a soft penalty and enforced during final thresholding

## Testing

Tests run for both numpy and torch backends. Coverage includes:

- acyclicity constraint is enforced
- correct extraction of DAG edges
- validation for invalid data types for logistic and poisson
- output matches reference implementation from xunzheng
- enforcement of forbidden, temporal, and required constraints

To run locally:

```bash
pre-commit run --all-files
pytest pgmpy/tests/test_causal_discovery/test_NOTEARS.py
```

## Trade offs

- torch backend helps on large datasets but adds overhead for small ones
- required edges are enforced at the end so user provided cyclic required edges can conflict with acyclicity
- reference numpy implementation is copied into tests instead of adding a dependency

---

# The following checklist is mandatory.

Your PR will be closed if you remove the checklist or do not answer the questions to a satisfactory level. Use of LLM is strictly forbidden for any part of this checklist (even for improving language).

### Your checklist for this pull request

- [x] Have you followed all the steps from our Contributing Guide?
- [x] Does the PR fully address the linked issue and is within its defined scope?
- [x] Are all the GitHub Actions checks passing?

Please answer the following questions:

- Did you use an LLM for any assistance? Please describe how and what you used it for?

Yes.

Used an LLM for the initial draft of the implementation after reading the requirements and the NOTEARS paper. Then went through the code, aligned it with pgmpy patterns, and refactored it to fit the causal_discovery module and sklearn style API.

- What steps have you taken to verify that the changes correctly address the issue? And what edge cases have you considered?

Tests added for both numpy and torch backends. Coverage includes:

- acyclicity constraint behavior
- forbidden edge enforcement
- temporal order enforcement
- required edge handling
- output consistency with reference implementation
- loss functions l2, logistic, poisson

Input validation tests include:

- invalid loss type
- non numeric data
- logistic with non binary data
- poisson with negative data

Also ran tests for other estimators to check for regressions.

- Has the LLM added try except blocks? They will need to be removed. Any error handling must be explicit.

No try except blocks added. Validation uses explicit checks and raises ValueError where needed.

- Have you used LLM for generating tests? They need to be compressed into a smaller number of tests without reducing coverage.

No.

### Issue number(s) that this pull request fixes

- Fixes #1342

### List of changes to the codebase in this pull request

- Add NOTEARS under `pgmpy/causal_discovery`
- Add tests under `pgmpy/tests/test_causal_discovery/test_NOTEARS.py`
- Add sklearn compatible API integration
- Add support for l2, logistic, and poisson loss
- Add constraint handling for forbidden, temporal, and required edges
- Add reference comparison in tests

### CODE REVIEW

**Review by DARHWOLF — CHANGES_REQUESTED** [review]
I see em dashes in the PR descriptions which indicates to me LLM usage. This is strictly forbidden as the PR template mentions. 
- I would like to see an entirely self written PR description. This will not be reviewed until then. Marking as AI Overuse.

- Has there been a design discussion on this?

**Comment by DARHWOLF** [comment]
I also see a large number of lines of code with very few commits. This is very difficult without LLM assistance. 
The PR description however says no LLM assistance has been used.

**Comment by GunaPalanivel** [comment]
> I see em dashes in the PR descriptions which indicates to me LLM usage. This is strictly forbidden as the PR template mentions.
> 
> * I would like to see an entirely self written PR description. This will not be reviewed until then. Marking as AI Overuse.
> * Has there been a design discussion on this?

I have rewritten the PR description.

For design discussion, there was prior discussion here: https://github.com/pgmpy/pgmpy/issues/2513#issuecomment-3968965229

**Comment by GunaPalanivel** [comment]
> I also see a large number of lines of code with very few commits. This is very difficult without LLM assistance. The PR description however says no LLM assistance has been used.

Regarding the commit size: most of the work was implemented locally before opening the PR, which is why it appears as a small number of commits. If it helps review, I can split the changes into smaller commits.

The implementation in this PR follows that direction and the structure used by existing estimators in pgmpy.

Let me know if you would prefer the PR to be broken into smaller commits for easier review.

cc: @DARHWOLF

**Comment by GunaPalanivel** [comment]
@ankurankan  Appreciate your review here :)

**Comment by ankurankan** [comment]
@GunaPalanivel 
1. The algorithm needs to go to `causal_discovery` module and the implementation needs to be sklearn compatible.
2. The tests need to be compressed. They look LLM generated.
3. Ideally, we need to compare our implementation with some other package that implements the algorithm and see if the values match.
4. Changes in docs need to be reverted.

I would like to review this PR on a call. Once you are done with the changes, can you message me on Discord to figure out a time?

**Review by ankurankan — CHANGES_REQUESTED** [review]
Refer comment

**Comment by GunaPalanivel** [comment]
> Refer comment

Hey @ankurankan, thanks for the review! I've pushed the changes:

- Moved NOTEARS to `causal_discovery` module, now inherits from `_BaseCausalDiscovery` (sklearn compatible - passes `parametrize_with_checks`) 
- Compressed the tests, organized them like `test_GES.py` with class-based structure
- Reverted all doc changes

Will message you on Discord to set up a time for the review call.
